_SPIKE_.md

```diff
@@ -0,0 +1,496 @@
+# Spike: Resolve included models
+
+Consider the following domain model(s):
+
+- A model called `Category` with properties `id`, `name`.
+- A model called `Product` with properties `id`, `name`, `categoryId`
+- A `hasMany` relation (Category has many products)
+
+Now consider the following code to retrieve all categories with all related
+products (perhaps to render a home page of a product catalog):
+
+```ts
+categoryRepo.find({include: [{relation: 'products'}]});
+```
+
+## The problem
+
+How to fetch products for each category found?
+
+Additional constraints:
+
+- The target model (`Product`) can be backed by a different database than the
+  source model (`Category`). For example, we can use MongoDB to store categories
+  and MySQL to store products.
+
+- We need to test relations against real databases to ensure we are correctly
+  handling database-specific quirks like:
+  - a limit on the number of items that can be included in `inq`
+  - coercion between `ObjectID` vs. `string` for MongoDB.
+
+## Proposed solution
+
+### 1. Introduce the concept of `InclusionResolver`
```

Comment:
+1 to have `InclusionResolver` so that we can build up virtual properties, such as fetching it from a REST api call or a database.

Reply:
Maybe the resolution can be handled as follows:

1. Use a registered resolver if exists
2. Check the named property/method on the entity
  - a method - calling the method to resolve the virtual property
  - a property (including getter) - use the value

Reply:
> Check the named property/method on the entity

That does not make sense to me. A model/entity class represents the data structure, it does not know how to store or load that data. Data access is implemented by Repositories.

For the initial version, let's keep the scope small (think of MVP) and implement only the first step - "Use a registered resolver if exists". We can look into different ways (like using a named property or method) later, when we have a more concrete requirement for that.

Reply:
Fair enough.